### PR TITLE
Add follows back helper

### PR DESCRIPTION
**Pull request by sbpipb**
This PR is for the [ticket 22](https://github.com/enderahmetyurt/gistcatch/issues/22)

This PR adds a "Follows you" label for the users that follow you back.

screenshot from 2018-10-24 20-50-31

Add following label for followers that you follow

- [x] validates if user follows the authenticated user
- [x] adds label "Follows you"

### CODE REVIEW

**Comment by sbpipb** [comment]
I have problems setting my frontend working, after changing the git tokens,  clicking sign in redirects to a dead page.

screenshot from 2018-10-23 23-55-57


This is the URL after clicking
`https://github.com/login/oauth/authorize?client_id&redirect_uri=http%3A%2F%2F127.0.0.1%3A3000%2Fusers%2Fauth%2Fgithub%2Fcallback&response_type=code&scope=user%3Aemail%2Cgist&state=42f84959a23ec868a77390f5b0705db44f845486d805b2e1`

I am keen on finishing this PR, cause it looks like it just needs some CSS changes

**Comment by enderahmetyurt** [comment]
You need to create an app for the development environment and set callback URL as localhost:3000 or where your rails server working on. It's all about your local settings.

**Comment by shivanibhanwal** [comment]
@sbpipb you can use these icon for the Following 
https://fontawesome.com/v4.7.0/icon/link
Or
https://fontawesome.com/v4.7.0/icon/retweet
and when the user opens the Gist page for his follower then the user can see on the top right 
`icon with following text`
and on dashboard then you can have this icon on the Bottom-right instead of center
cc @enderahmetyurt

**Comment by enderahmetyurt** [comment]
I guess Twitter does not show any icon to represent the following status. We can do it the same way. @sbpipb your first image is enough for me. I think we don't need an icon for that. We can put just follows text label under Github login name. By the way, that label should appear only following user cards. If the user who I follow doesn't follow me, we can put "-" or something under the Github login name. @sbpipb could you develop that issue in this way and ping me when you have done. I will check and let's discuss it. Cheers.

**Comment by sbpipb** [comment]
> By the way, that label should appear only following user cards. If the user who I follow doesn't follow me, we can put "-" or something under the Github login name.

I don't think putting a "-" is appropriate for users who the current user doesn't follow. It doesn't look good, or intuitive, ideally the other websites use a button (CTA) to indicate if user is followed or not.

**Comment by enderahmetyurt** [comment]
Yes, you're right "-" doesn't look good but if the user doesn't follow me, what kind of button do you put on there?

**Comment by sbpipb** [comment]
> Yes, you're right "-" doesn't look good but if the user doesn't follow me, what kind of button do you put on there?

Ideally, it would be a follow button, when the user clicks this, it would lead to Github page.

Maybe a link similar to this one?
https://buttons.github.io

**Comment by enderahmetyurt** [comment]
No, you cannot follow button there because you already follow the user. The user doesn't follow you. We want to put it following page not followers page. You follow for example X, X doesn't follow you. What could you do for that? What kind of button can you put? Follow? You already follow X. Got it?

**Comment by sbpipb** [comment]
Hi, please review the requirements, it says this 
> On follower page, if the user follows authenticated user, shows the following icon otherwise show nothing.

So, I should be putting a label only for the list of users that follow me, not the users I follow.  If this is not the requirements, please update it.

**Comment by sbpipb** [comment]
I get what you mean now, but the requirements has a typo, it should have been 

On the `following` page, if the user follows authenticated user, shows the following icon otherwise show nothing.

**Comment by sbpipb** [comment]
Done updating the code, as well as PR documentation

**Comment by enderahmetyurt** [comment]
but in the image, you put first comment. Do you see the wrong? I meant, if the user doesn't follow you, what can we put there? If we blank as you did, it looks weird. Please do it proper way :(

**Comment by sbpipb** [comment]
let's go back a bit.

Firstly, which user list should be affected by this "label" change, the `Followers` list or `Following` list?

**Comment by sbpipb** [comment]
< No, you cannot put a follow button there because you already follow the user. The user doesn't follow you. We want to put it following page not followers page. You follow for example X, X doesn't follow you. What could you do for that? What kind of button can you put? Follow? You already follow X. Got it?


If the user doesn't follow you, then we leave is blank,  but for the html we keep the same height so it looks uniformed.. what do you say?

screenshot from 2018-10-24 21-40-22

**Comment by enderahmetyurt** [comment]
YES! what I meant. Now there is div space. What do we put there?

File: app/views/dashboard/index.html.erb [hunk 1]
```diff
@@ -19,6 +19,7 @@
             <img class="card-img-top" src="<%= f.avatar_url %>" alt="Card image cap">
             <div class="card-block">
               <span class="card-title"><%= f.login %></span>
+                <div class='card-status'><%= following?(f, @followers) ? "Follows you" : "&nbsp;"%></div>
```
Comment (enderahmetyurt) [inline app/views/dashboard/index.html.erb]: f is object here.

File: app/helpers/dashboard_helper.rb [hunk 2]
```diff
@@ -26,4 +26,11 @@ def gist_action(action)
         }
       end
   end
+
+  # checks a particular user if is being followed
+  # given a follower list
+  # much faster than doing client.follows?
+  def following?(user_id, following_list)
+    following_list.any? { |f| user_id == f.id }
```
Comment (enderahmetyurt) [inline app/helpers/dashboard_helper.rb]: it should be { |f| user_id.id == f.id }

File: app/views/dashboard/index.html.erb (hunk 1 above)
Comment (enderahmetyurt) [reply app/views/dashboard/index.html.erb]: Also "&nbsp;" doesn't put a space. It puts ""&nbsp;" string.

**Comment by enderahmetyurt** [comment]
I do some comments and there is a style bug. It doesn't look like that. Please fix and test before open a PR.
[Screenshot](https://ibb.co/coEzOA)

PS: Sorry for the screenshot. I couldn't upload the image to Github.

**Comment by enderahmetyurt** [comment]
by the way. @sbpipb please rebase the master your PR branch. Because there are big changes about views and methods.

**Comment by enderahmetyurt** [comment]
This is your code how it's working
screen shot 2018-10-24 at 21 58 51

**Comment by sbpipb** [comment]
Will send my updates in an hour

File: app/views/dashboard/index.html.erb [hunk 3]
```diff
@@ -28,17 +31,15 @@
...
-    <div class="container">
-      <% @followers.each do |f| %>
-        <%= link_to get_gists_path(f.login) do %>
-          <div class="card bg-secondary mb-3 stampsize follower">
-            <img class="card-img-top" src="<%= f.avatar_url %>" alt="Card image cap">
-            <div class="card-block">
-              <span class="card-title"><%= f.login %></span>
-            </div>
+    <% @followers.each do |f| %>
+      <%= link_to get_follower_gists_path(f.login) do %>
```
Comment (enderahmetyurt) [inline app/views/dashboard/index.html.erb]: is it a proper route? Please fix it

File: app/views/dashboard/index.html.erb [hunk 4]
```diff
@@ -28,17 +31,15 @@
...
+        <div class="card bg-secondary mb-3 stampsize follower">
+          <img class="card-img-top" src="<%= f.avatar_url %>" alt="Card image cap">
+          <div class="card-block">
+            <span class="card-title"><%= f.login %></span>
           </div>
-        <% end %>
+        </div>
       <% end %>
```
Comment (enderahmetyurt) [inline app/views/dashboard/index.html.erb]: Missing and why?

File: app/views/dashboard/index.html.erb [hunk 5]
```diff
@@ -19,6 +19,9 @@
             <img class="card-img-top" src="<%= f.avatar_url %>" alt="Card image cap">
             <div class="card-block">
               <span class="card-title"><%= f.login %></span>
+              <div class='card-status'>
+                <%= following?(f.id, @followers) ? "Follows you" : "&nbsp;"%>
```
Comment (enderahmetyurt) [inline app/views/dashboard/index.html.erb]: it should be "&nbsp;".html_safe. Please fix and test please :(

**Comment by enderahmetyurt** [comment]
Before creating a PR please test all app. I found a few errors and write my comments.

**Comment by enderahmetyurt** [comment]
Or nevermind. I will merge it and I will fix it on master branch your code which you missed. Thank you for contribution.

**Pull request merged**